text/0000-system-receptionist.md

```diff
@@ -0,0 +1,256 @@
+
+The `system` value is made available in the 3rd argument to action/guard/invoke creators, which contains the `system.receptionist` that extends the `ActorRef` interface.
+
+Listing all `ActorRef` objects registered under a key is done via `system.receptionist.list(key)`, which synchronously returns an array of `ActorRef` objects.
+
+```ts
+const todosMachine = createMachine({
+  // ...
+  on: {
+    'todos.logAll': {
+      actions: (ctx, e, { system }) => {
+        // Lists all todo ActorRefs
+        console.log(system.receptionist.list('todo'));
+      }
+    },
+    'todos.completeAll': {
+      actions: pure((ctx, e, { system }) => {
+        return system.receptionist.list('todo').map(todoRef => {
+          return sendTo(todoRef, { type: 'complete' });
+        });
+      });
+    }
+  }
+});
+```
+
+This should be roughly equivalent to:
+```ts
+// PSEUDOCODE
+// system.receptionist.list('todo');
+system.receptionist.getSnapshot().registry.get('todo') ?? []; // => ActorRef[]
+```
+
+**Single listing**
+
+A common use-case of the receptionist will be to register and find a single `ActorRef` registered to a given `key`. This can be done via `system.registry.find(key)`, which will return `ActorRef | null`.
+
+```ts
+const todoMachine = createMachine({
+  // ...
+  on: {
+    complete: {
+      // `system.registry.find(key)` returns 0 or 1 ActorRef objects
+      // TODO: ensure `sendTo(...)` handles null targets (dead-letter queue?)
+      actions: sendTo((ctx, e, { system }) => system.receptionist.find('notifier'), {
+        type: 'notify',
+        message: 'Todo completed'
+      })
+    }
+  }
+});
+```
+
+This should be roughly equivalent to:
+```ts
+// PSEUDOCODE
+// system.receptionist.get('notifier');
+system.receptionist.getSnapshot().registry.get('notifier')?.[0] ?? undefined; // => ActorRef[]
+```
+
+**Subscribing to listings**
+
+The registry may change over time, as actors get registered and unregistered from certain keys. The `system.receptionist.listener(key)` returns an `ActorRef` (subscribable by nature) which can be subscribed to for the specific purpose of listening to changes in `ActorRef` objects registered under a key. The emitted value is the complete `ActorRef[]` array.
+
+```ts
+const todosMachine = createMachine({
+  invoke: {
+    src: (ctx, e, { system }) => fromObservable(() => system.receptionist.listener('todo')),
+    onEmit: {
+      actions: (_, event) => {
+        // ActorRef[] of 'todo' actor refs
+        console.log(event.data); 
+      }
+    }
+  }
+});
+```
+
+This should be roughly equivalent to:
+```ts
+// PSEUDOCODE
+// system.receptionist.listener('todo')
+const receptionist = {
+  onRegister(key, actorRef) {
+    // Register actorRef
+    // ...
+
+    observers[key].forEach(observer => {
+      observer.next(registry.get(key))
+    });
+  },
+
+  listener(key) {
+    // Return ActorRef that listens for registry updates on that key
+    // ...
+  }
+}
+```
+
+
+**Unregistering**
```

Comment:
I have no idea what to do about it yet but it's worth noting that a code like this isn't guaranteed to work in the React world:
```ts
// one component
useMachine({
  entry: register('todo')
})

// another component
useSelector(system.registry.find('todo'))
```

The problem is with the timing of the `register` - we can't easily call it within the render function and thus we might not be able to read from that registered key in another part of the app.

Perhaps the upcoming cache API could help us to solve this but there are not a lot of details available out there about this so I'm not sure.

Reply:
@Andarist See the latest changes, where registering is done by the _parent_ at the time of actor creation, rather than as an action on the child:


```ts
import { register } from 'xstate/actions';
const todoMachine = createMachine({
  // Register actor to the 'todo' key
  entry: register('todo'),
  // ...
});
```

This should be roughly equivalent to:
```ts
// PSEUDOCODE
entry: (ctx, e, { system, self }) => {
  system.register('todo', self);
}
```

Reply:
```ts
const todoMachine = createMachine({
  // ...
});

const todosMachine = createMachine({
  on: {
    'todo.create': {
      actions: assign({
        newTodo: (ctx, e, { spawn }) => {
          return spawn(todoMachine, {
            id: `todo-${/* unique id */}`,
            register: 'todo' // multiple actors can be registered here
          })
        }
      })
    }
  }
})
```

Latest changes; registering now occurs at time of creation

Reply:
that doesn't change anything for the case like this:
```ts
createMachine({
  context: ({ spawn }) => spawn(todoMachine, { register: 'todo' })
})
```

The problem is still the same - the registry can't be easily mutated during render and thus it can't be reliably read during the same render by another component.

Reply:
In the above case, the spawned machine is available immediately; that is, it's part of `machine.getInitialState(...)` and doesn't need to wait to be started in an effect, since the instance will be the same (just not started yet).

Reply:
Yes, this part is OK - the problem is with the "reading" part here:
```ts
useSelector(system.registry.find('todo'))
```
We can't reliably make the `'todo'` key available during the same render without introducing potential memory leaks in the React model. Some libraries that try to solve this are using **global** timers and/or `FinalizationRegistry`... but that's a giant hack in itself.

The problem is that the "key owner" (`todosMachine` in your example) might not end up mounting despite being rendered at least once - in such a case the key itself should become unavailable in the registry... but how do you cleanup the key added in render?

Reply:
The `'todo'` key would be available. When the system is started (machine is started), `system.registry.find('todo')` is immediately available.

Are you approaching this from the concern of components _externally_ registering actors onto the registry? This is not supported/mentioned in this RFC, as only actors within the system can register other actors.

Reply:
```ts
const Child = () => {
  // the key might not be available here yet
  // even though its parent component "registers" this key at the creation time
  // and not when the machine in the parent starts
  useSelector(system.registry.find('todo')) 
}

const Parent = () => {
  // not that this key wants to conceptually be available asap as it's registered at creation time
  // if we make it available immediately in the registry (that is a shared resource) this would be equivalent to a side-effect in render
  // it's not possible, in a clean, predictable, way to remove such a key in case the `<Parent/>` doesn't get committed by React
  createMachine({
    context: ({ spawn }) => spawn(todoMachine, { register: 'todo' })
  })
  return <Child/>
}
```

Reply:
There's a slight misunderstanding: `system` is not "global".

This would look more like this:

```js
const Child = ({ mySystem }) => {
  // Available
  useSelector(mySystem.registry.find('todo'));
}

const Parent = () => {
  const mySystem = useInterpret(createSystem(...)); // system is just an actor

  return <Child mySystem={mySystem} />
}
```

Reply:
It doesn't matter if the system is truly global or if it's scoped to a React subtree - it's still a shared resource.

There is still a scenario in which the system actor gets created and committed while a machine registering into it from within its subtree won't do the same. Imagine a conditional rendering within the subtree, you might render `ChildA` (registering `todo-a`) but before it commits you can switch to `ChildB` (registering `todo-b`). In such a case, it's not obvious how to clean the `todo-a` key from the system

Reply:
With the latest changes to this RFC, a machine _never_ registers itself to a system. It's the parent machine that registers child machines to the system.

Contrived example: adding a todo actor (via sending a `'todo.create'` or similar event to the parent actor) would do these things:

1. Add it to the machine context via `spawn(...)`
2. Simultaneously register it with the `system` that the parent actor belongs to
3. **React will rerender**
  - At this point, that todo actor exists.

Note that React is not involved _at all_ until after the actor is registered.